heron/dsl/src/java/com/twitter/heron/dsl/impl/BuilderImpl.java

```diff
@@ -0,0 +1,58 @@
+//  Copyright 2017 Twitter. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package com.twitter.heron.dsl.impl;
+
+
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+import com.twitter.heron.api.topology.TopologyBuilder;
+import com.twitter.heron.dsl.Builder;
+import com.twitter.heron.dsl.Streamlet;
+
+/**
+ * BuilderImpl implements the Builder interface.
+ * The builder keeps track of the all the sources of the
+ * computation dag and builds the Topology by traversing all
+ * the computation nodes.
+ */
+public final class BuilderImpl implements Builder {
+  private List<StreamletImpl<?>> sources;
```

Comment:
why StreamletImpl and not just Streamlet?

Reply:
Because BuilderImpl needs access to doBuild which is only in StreamletImpl

Reply:
Renamed StreamletImpl to BaseStreamlet.
Did same for KV